### PR TITLE
[Cisco ASA] Fix GROK to add support for spaces and special characters in usernames

### PR DESCRIPTION
**Pull request by chemamartinez**
## Proposed commit message

- **fb791cae29db3dd2c576ec5b6d999dc6191d600d**

Adds support for spaces and special characters in usernames. According to the [Cisco docs](https://www.cisco.com/c/en/us/td/docs/wireless/wlse/2-8/user/guide/app_name.html#wp1011885) there are some characters that are allowed to be user for usernames and are not taken into account.

Some events delimite usernames with spaces, so for now the PR covers usernames that are between parentheses.

- **acf8f2d05301b5d6d3dfbe6bdac9429b885cb054**

It fixes the parse of this event:
```
<140>Oct 06 2023 10:11:00 myAsaHostname : %ASA-4-313005: No matching connection for ICMP error message: icmp src outside:192.168.2.2(9999:my_SgtNameSrc) dst inside:192.168.2.3(8888:my_SgtNameDst) (type 3, code 3) on outside interface.  Original IP payload: udp src 192.168.2.3/53 dst 192.168.2.2/60919.
```

Where the SGT tag can appear for the destination as well as the source.

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

Added pipeline tests to cover new cases:
```console
# elastic-package test pipeline --generate
Run pipeline tests for the package
--- Test results for package: cisco_asa - START ---
╭───────────┬─────────────┬───────────┬──────────────────────────────┬────────┬──────────────╮
│ PACKAGE   │ DATA STREAM │ TEST TYPE │ TEST NAME                    │ RESULT │ TIME ELAPSED │
├───────────┼─────────────┼───────────┼──────────────────────────────┼────────┼──────────────┤
│ cisco_asa │ log         │ pipeline  │ test-additional-messages.log │ PASS   │     18.203ms │
│ cisco_asa │ log         │ pipeline  │ test-anyconnect-messages.log │ PASS   │   7.588167ms │
│ cisco_asa │ log         │ pipeline  │ test-asa-fix.log             │ PASS   │   9.515958ms │
│ cisco_asa │ log         │ pipeline  │ test-asa-missing-groups.log  │ PASS   │   8.462125ms │
│ cisco_asa │ log         │ pipeline  │ test-asa.log                 │ PASS   │  84.864375ms │
│ cisco_asa │ log         │ pipeline  │ test-dap-records.log         │ PASS   │   4.271375ms │
│ cisco_asa │ log         │ pipeline  │ test-filtered.log            │ PASS   │   4.044916ms │
│ cisco_asa │ log         │ pipeline  │ test-hostnames.log           │ PASS   │   4.407083ms │
│ cisco_asa │ log         │ pipeline  │ test-non-canonical.log       │ PASS   │  10.056292ms │
│ cisco_asa │ log         │ pipeline  │ test-not-ip.log              │ PASS   │   5.195292ms │
│ cisco_asa │ log         │ pipeline  │ test-sample.log              │ PASS   │  42.080083ms │
│ cisco_asa │ log         │ pipeline  │ test-sgt-tag-name.log        │ PASS   │   18.17725ms │
│ cisco_asa │ log         │ pipeline  │ test-sip.log                 │ PASS   │   5.741125ms │
╰───────────┴─────────────┴───────────┴──────────────────────────────┴────────┴──────────────╯
--- Test results for package: cisco_asa - END   ---
Done
```

### CODE REVIEW

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 1]
```diff
@@ -393,7 +393,7 @@ processors:
         IPORHOST: "(?:%{IP}|%{HOSTNAME})"
         NOTCOLON: "[^:]*"
         CISCO_USER_OR_SGT: (?:%{CISCO_USER}|%{CISCO_SGT})
-        CISCO_USER: (?:\*\*\*\*\*|(?:(?:LOCAL\\)?(?:%{HOSTNAME}\\)?%{USERNAME}\$?(?:@%{HOSTNAME})?%{CISCO_SGT}?))
+        CISCO_USER: (?:\*\*\*\*\*|(?:(?:LOCAL\\)?(?:%{HOSTNAME}\\)?%{USERNAME}\$?(?:@%{HOSTNAME})?%{CISCO_SGT}?)|[^$)]+)
```
Comment (efd6) [inline packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:396]: Can you explain the intention behind this pattern?

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 2]
```diff
@@ -1792,6 +1792,19 @@ processors:
   - remove: 
       field: _temp_.cisco.destination_username
       if: ctx._temp_?.cisco?.destination_username != null && ctx._temp_.cisco.destination_username == ""
+  # Trim $ from source_username, destination_username
```
Comment (efd6) [inline packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:1795]: Why is this necessary?

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 1 above)
Comment (chemamartinez) [reply packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:396]: For that specific log format we know the username is between parentheses, so the intention is to catch everything inside. I also excluded `$` because I saw in https://github.com/elastic/integrations/pull/8362 that when an username ends by `$`, it should not be considered part of the field.

Also, I added the change at the end of the pattern attempting to make the less impact as possible in the current patterns.

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 2 above)
Comment (chemamartinez) [reply packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:1795]: It was the easiest way I found to fix the parsing of `(LOCAL\\myUser1234$, 9999:my_SgtName)`. For some reason, when the username contains the `$`, the SGT was not being parsed. After the changes of my first commit, it was being parsed but `$` was being included at the end of the username and I couldn't find a way to remove it in the original pattern.

Definitely it should be a cleaner way to achieve this, I'll give it another iteration.

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 1 above)
Comment (efd6) [reply packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:396]: In general, if I'm matching on something that is bracketed, I'll use an alternation including the possible bracketing states; so one branch with brackets and one brance without `(?:unbracketed|\(bracketed\))`.

Comment (chemamartinez) [reply packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:396]: Brackets are actually in the main pattern, outside the `%CISCO_USER` one. However, if I let the `)` character not be a limit to end match, it includes it in the username.

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 2 above)
Comment (chemamartinez) [reply packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml]: 43f2930b3cdce7cd73c5b7f4f8b9ab575d2a2fed fixes this case by matching the `$` outside the capture group for the user, with the same desired result.

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 1 above)
Comment (efd6) [reply packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:396]: This passes.
```diff
diff --git a/packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml b/packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml
index 494f46c85..2aa7a4e2b 100644
--- a/packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml
+++ b/packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml
@@ -387,12 +387,12 @@ processors:
       field: "message"
       description: "106023"
       patterns:
-        - ^%{NOTSPACE:event.outcome} ((protocol %{POSINT:network.iana_number})|%{NOTSPACE:network.transport}) src %{NOTCOLON:_temp_.cisco.source_interface}:%{IPORHOST:source.address}(/%{NUMBER:source.port})?\s*(\(%{CISCO_USER_OR_SGT:_temp_.cisco.source_user_or_sgt}\) )?dst %{NOTCOLON:_temp_.cisco.destination_interface}:%{IPORHOST:destination.address}(/%{NUMBER:destination.port})?%{DATA}by access-group "%{NOTSPACE:_temp_.cisco.list_id}"
+        - ^%{NOTSPACE:event.outcome} ((protocol %{POSINT:network.iana_number})|%{NOTSPACE:network.transport}) src %{NOTCOLON:_temp_.cisco.source_interface}:%{IPORHOST:source.address}(/%{NUMBER:source.port})?\s*(\(%{CISCO_USER_OR_SGT}\) )?dst %{NOTCOLON:_temp_.cisco.destination_interface}:%{IPORHOST:destination.address}(/%{NUMBER:destination.port})?%{DATA}by access-group "%{NOTSPACE:_temp_.cisco.list_id}"
       pattern_definitions:
         HOSTNAME: "\\b(?:[0-9A-Za-z][0-9A-Za-z-_]{0,62})(?:\\.(?:[0-9A-Za-z][0-9A-Za-z-_]{0,62}))*(\\.?|\\b)"
         IPORHOST: "(?:%{IP}|%{HOSTNAME})"
         NOTCOLON: "[^:]*"
-        CISCO_USER_OR_SGT: (?:%{CISCO_USER}|%{CISCO_SGT})
+        CISCO_USER_OR_SGT: (?:%{CISCO_USER:_temp_.cisco.source_user_or_sgt}|%{CISCO_SGT:_temp_.cisco.source_user_or_sgt}|\((?:%{CISCO_USER:_temp_.cisco.source_user_or_sgt}|%{CISCO_SGT:_temp_.cisco.source_user_or_sgt})\))
         CISCO_USER: (?:\*\*\*\*\*|(?:(?:LOCAL\\)?(?:%{HOSTNAME}\\)?%{USERNAME}\$?(?:@%{HOSTNAME})?%{CISCO_SGT}?)|[^$)]+)
         CISCO_SGT: (?:, *)?%{NUMBER}(?::%{WORD})?
   - dissect: 
```

Comment (chemamartinez) [reply packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:396]: I've tried those suggestions but it still gives problems when brackets are optional, for example in event 302014, where the last bracket is still included in the username following that pattern. I've created a [tmp branch](https://github.com/chemamartinez/integrations/commit/6929a1478ca9fe460c59b57d2af259a7720299c5) in case you'd want to take a look at the change.

Comment (efd6) [reply packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:396]: The issue is [here](https://github.com/chemamartinez/integrations/blob/6929a1478ca9fe460c59b57d2af259a7720299c5/packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml#L1710-L1729) and by parallel likely also in the following groks.

Comment (chemamartinez) [reply packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml]: In that case, I've updated the commit with the suggested changes and excluded the bracket from the username in the grok you pointed to. 87891898ec268f4ee87f22cc5c5097ad2797bc88

**Review by efd6 — APPROVED** [review]
